src/python/gudhi/clustering/_tomato.cc

```diff
@@ -243,56 +249,55 @@ auto merge(py::array_t<Cluster_index, py::array::c_style> children, Cluster_inde
   for (Cluster_index j = 0; j < n_leaves; ++j) {
     Cluster_index k = ds.find_set(j);
     if (ds_bas[k].name == -1) ds_bas[k].name = next_cluster_name++;
-    ret.push_back(ds_bas[k].name);
+    ret->push_back(ds_bas[k].name);
   }
-  return py::array(ret.size(), ret.data());
+  return _wrap_as_numpy_array(ret, ret->size());
 }
 
 // TODO: Do a special version when ngb is a numpy array, where we can cast to int[k][n] ?
-// py::isinstance<py::array_t<std::int32_t>> (ou py::isinstance<py::array> et tester dtype) et flags&c_style
-// ou overload (en virant forcecast?)
-// aussi le faire au cas où on n'aurait pas un tableau, mais où chaque liste de voisins serait un tableau ?
-auto hierarchy(py::handle ngb, py::array_t<double, py::array::c_style | py::array::forcecast> density)
+// also do this in the case where we don't have an array, but each list of neighbours is an array ?
+auto hierarchy(nb::object ngb, nb::ndarray<const double, nb::ndim<1> > density)
 {
   // used to be py::iterable ngb, but that's inconvenient if it doesn't come pre-sorted
   // use py::handle and check if [] (aka __getitem__) works? But then we need to build an object to pass it to []
   // (I _think_ handle is ok and we don't need object here)
-  py::buffer_info wbuf = density.request();
-  if (wbuf.ndim != 1) throw std::runtime_error("density must be 1D");
-  const int n = wbuf.shape[0];
-  double* d = (double*)wbuf.ptr;
+  const int n = density.shape(0);
   // Vector { 0, 1, ..., n-1 }
   std::vector<Point_index> order(boost::counting_iterator<Point_index>(0), boost::counting_iterator<Point_index>(n));
+  auto density_view = density.view();
   // Permutation of the indices to get points in decreasing order of density
-  std::sort(std::begin(order), std::end(order), [=](Point_index i, Point_index j) { return d[i] > d[j]; });
+  std::sort(std::begin(order), std::end(order), [&density_view](Point_index i, Point_index j) {
+    return density_view(i) > density_view(j);
+  });
   // Inverse permutation
   std::vector<Point_index> rorder(n);
   for (Point_index i : boost::irange(0, n)) rorder[order[i]] = i;
   // Used as:
   // order[i] is the index of the point with i-th largest density
   // rorder[i] is the rank of the i-th point in order of decreasing density
-  // TODO: put a wrapper on ngb and d so we don't need to pass (r)order (there is still the issue of reordering the
-  // output)
-  return tomato(n, ngb, d, order, rorder);
+  // TODO: put a wrapper on ngb and density so we don't need to pass (r)order (there is still the issue of reordering
+  // the output)
+  return tomato(n, ngb, density, order, rorder);
 }
 
-PYBIND11_MODULE(_tomato, m)
+NB_MODULE(_tomato_ext, m)
 {
+  m.attr("__license__") = "MIT";
   m.doc() = "Internals of tomato clustering";
-  m.def("hierarchy", &hierarchy, "does the clustering");
-  m.def("merge", &merge, "merge clusters");
+  m.def("_hierarchy", &hierarchy, nb::arg("ngb"), nb::arg("density"), "does the clustering");
```

Comment:
This is negligible here, but note that naming the arguments causes some overhead (https://nanobind.readthedocs.io/en/latest/functions.html#minimizing-binding-overheads) and it isn't very useful for an internal function (a comment would be as informative).

Reply:
Oh, I missed this part of the documentation. There are probably a lot other methods like that, I will take care of it.